Increment1-3

ncyellow opened the pull request:
Задание первого спринта

ncyellow committed:
Задание по инкременту 1 (пока нет конфигурирования хоста порта)

ncyellow committed:
пока набросок

ncyellow committed:
поменял ветки для тестов на increment2

ncyellow committed:
По результатам тестов убрал проверки на допустимость только указанных в задании тегов + расставил нужные http статусы

ncyellow committed:
Пофиксил один из тестов и добавил все тестовые сценарии инкремента 2

ncyellow committed:
Решив что я самый умный сделал проверку на Content-type из задачи. они ломала все тесты интеграционные + пара defer

ncyellow committed:
Узнал что такое statckcheck поправил замечания + попровил роут так как из-за него один из тестов падал

ncyellow committed:
фикс статической проверки

ncyellow committed:
Черепикнул коммит из increment1 который не попал в increment2

ncyellow committed:
Все кроме списка по корню. хочу посмотреть в каком формате ждет автотест списка метрик. пока метод просто возвращает рыбу

ncyellow committed:
Сделал формат вывода  .3f как в тесте

ncyellow committed:
стан ругается что не закрыто тело ответа

ncyellow committed:
Добавил вывод в html всех метрик + немного тестов и комментов

ncyellow committed:
не поправил тест на  html списка  метрик

ASPecherkin commented on cmd/server/main.go:
```diff
@@ -1,3 +1,26 @@
 
-func main() {}
+import (
+	"github.com/ncyellow/devops/internal/server/handlers"
```
просто на будущее - https://golangci-lint.run

лучше его поставить и настроить goimports , чтобы пакеты всегда были в одном порядке 

вида - сначала std , дальше внутренние пакеты и дальше внешние зависимости

ncyellow replied:
ok сделаю

ASPecherkin commented on internal/agent/agent.go:
```diff
@@ -0,0 +1,131 @@
+// sendToServer отправка метрик на сервер
+func (collector *Agent) sendToServer() {
+	//! приводим все метрики к нужным типам.
+	var data = map[string]float64{
```
так как метрики могут добавляться или убираться - я бы вынес создание этой map в отдельную функцию, в которую будет передан collector.metrics и на выходе уже map

ncyellow replied:
Без проблем

ASPecherkin commented on internal/server/handlers/handlers.go:
```diff
@@ -0,0 +1,115 @@
+}
+
+// ValueHandler обрабатывает GET запросы на чтение значения метрик. Пример /value/counter/counterName
+func ValueHandler(repo storage.Repository) http.HandlerFunc {
```
такой способ возможен, если аргументов немного 

Ещё один вариант - 
```go
type Handler struct {
 repo storage.Repository 
 log Logger 
...
}
```

и у него уже сделать метод, который вернёт http.HandlerFunc 

Это не замечение, скорее альтернативный подход

ncyellow replied:
Да я читал что есть второй вариант. Если аргументов станет много, то конечно надо будет поменять

ASPecherkin commented on internal/server/storage/memory.go:
```diff
@@ -0,0 +1,66 @@
+}
+
+func (s *MapRepository) UpdateCounter(name string, value int64) error {
+	s.counters[name] = s.counters[name] + value
```
тут есть добавление значение и вопрос - может быть ли оно конкуретным? 

если да - то может произойти паника при конкуретной записи 

но судя по коду выше - вызывается оно только в одном месте и здесь всё пройдёт. 

Информация для будущего 
Если я что-то упустил - поправь меня

ncyellow replied:
тут я не очень понял. Сейчас это дергается только из хендлера по обновлению. Ты имеешь ввиду что надо ставить Mutex? Я просто подумал раз этой темы в спринте не было то пока и делать не надо

ncyellow replied:
сделал защиту через RWMutex

ASPecherkin reviewed (changes requested):
есть небольшие замечения и комментарии - если есть вопросы по ним - можем обсудить

ncyellow commented:
> есть небольшие замечения и комментарии - если есть вопросы по ним - можем обсудить

Замечания понятны.
Я правильно понимаю мне чтобы пройти ревью надо добавить линтер + вынос в отдельную функцию. остальное просто для информации и описание альтернативных вариантов

ncyellow committed:
Исправлены замечания по ревью

1. Импорты по goimports
2. формирование map метрик в агенте в отдельных функциях
3. добавил RWMutex для защиты структуры данных от конкурентного доступа

ASPecherkin commented:
> > есть небольшие замечения и комментарии - если есть вопросы по ним - можем обсудить
> 
> Замечания понятны. Я правильно понимаю мне чтобы пройти ревью надо добавить линтер + вынос в отдельную функцию. остальное просто для информации и описание альтернативных вариантов

да, всё верно

ASPecherkin reviewed (approved):
Всё хорошо, работу принимаю
